### ISSUE TITLE
On Android 12 (One UI 4) FGA detects the wrong part of the screen

### ISSUE THREAD

**Issue by ScathachSkadi** [Fate-Grand-Automata/FGA#1077]
### Preparation

- [X] I tested the [latest release](https://github.com/Fate-Grand-Automata/FGA/releases/tag/canary)
- [X] I looked at other issues (even the [closed ones](https://github.com/Fate-Grand-Automata/FGA/issues?q=is%3Aissue+is%3Aclosed))
- [X] I read the [Troubleshooting Guide](https://github.com/MathewSachin/Fate-Grand-Automata/wiki/Troubleshooting)

### FGO server

JP, KR

### FGA build number

1599

### Describe the bug

My device recently got an Android 12 update, and it changed how some of the apps display on the screen.
It used to be either:
1. Has even black bars on both left and right sides
2. Only the camera part is back

But now it kind of does both. It is black on the camera part and it has a slim black bar on the other side. I used to use all the FateGO apps on full screen so I reapplied the full screen settings so it won't have any problems with FGA and I toggled the notch settings on. However, no matter what combinations I do,(making FGO full screen/non full screen, FGA app full screen/non full screen and notch toggle on/off) FGA will detect the wrong parts of the screen. I couldn't record with the FGA screen record because it kept crashing, so I used the native one but it still shows the debug mode.

This is how it looks now
![KakaoTalk_20220118_100940518](https://user-images.githubusercontent.com/70437676/149856165-6e9fb47a-594e-4c25-b29f-e471c651cfcc.jpg)
![KakaoTalk_20220118_104131643](https://user-images.githubusercontent.com/70437676/149856174-9a94e882-b7f4-4496-a486-163690a11c19.jpg)
![KakaoTalk_20220118_105136216_01](https://user-images.githubusercontent.com/70437676/149857007-786cf944-2018-4d76-ba22-e135c6f2c93b.jpg)
![KakaoTalk_20220118_105136216](https://user-images.githubusercontent.com/70437676/149857012-018c3996-6156-49ef-85b0-633e37e9a679.jpg)

I don't know if this is a Samsung thing or if it applies to all Android 12 devices,

### Video

https://streamable.com/03j6vw

FGO full screen mode with notch toggle on

https://streamable.com/plcd5m

FGO non full screen mode with notch toggle off

Let me know if you need more setting combinations.

### Device model

Galaxy S20+

### Android version

12

### Screen size

2400x1080

### RAM

12GB

**Comment by ScathachSkadi** [comment]
I tried the custom game area mode but since I don't know which parts have to match I couldn't get it to work :(

**Comment by raahilsha** [comment]
Hi! I had this problem on my Fold 3 on Android 12, as described in this issue: https://github.com/Fate-Grand-Automata/FGA/issues/1072

You can see how I resolved it here: https://github.com/Fate-Grand-Automata/FGA/issues/1072#issuecomment-1018614258

There's also an option to change how FGO interacts with the camera cutout, and you can try forcing "Show Camera Cutout" to get the Android 11 display mode back as well.

**Comment by pristine-estimate** [comment]
I had this same problem on a Fold 3 running Android 12 as described in this issue: https://github.com/Fate-Grand-Automata/FGA/issues/1072

and was able to resolve it as in this comment: https://github.com/Fate-Grand-Automata/FGA/issues/1072#issuecomment-1018618429

There's also an option to change the camera cutout behaviour, and you can force it to "Show Camera Cutout" to get the game to display as it did on Android 11.

**Comment by ScathachSkadi** [comment]
I've been messing around with that option for days. And the problem is that it still doesn't work. Ever since the new ratio update in FGO, making it fullscreen solved all the issues. However, now, it somehow detects the wrong parts of the screen even when it looks and behaves the same as before. Also, I tried the show camera cutout option as well. And I toggled the notch option both on and off when I tested. Thanks for your reply, but I guess it doesn't work for me.

**Comment by tofu951753** [comment]
Also joining in to get updated on this issue. I was playing around with it and having fga on custom area with left +38 worked for me to an extent, but it can't select supports after since the detection will be off. 

Note 10, Android 12, FGA 1611 (Don't remember previous build but I updated to latest since I thought it was a bug)

Let me know if you find a fix @ScathachSkadi

**Comment by reconman** [comment]
@tofu951753 Support screen is the most important screen, battle is a bit more tolerant. Can you provide a screenshot of the support selection screen where the red rectangles are visible?

@ScathachSkadi All the red rectangles are fine but it seems like FGA doesn't see anything anymore. I have the same display ratio and the same FGO button locations.

Maybe MediaProjection is broken on some Android 12 devices?

**Comment by ScathachSkadi** [comment]
@reconman Is there a way to find out if the MediaProjection is broken? Some kind of log export? I didn't try reinstalling the whole app yet, and I had FGA installed for months. Would reinstalling might help?

**Comment by reconman** [comment]
> I couldn't record with the FGA screen record because it kept crashing, so I used the native one but it still shows the debug mode.

This is the problem. Can you try pulling some crash logs by using `adb logcat -E > log.txt` on a PC?

**Comment by ScathachSkadi** [comment]
Ah, right. The logcat. Sadly, I won't have access to a PC for a while. At least, I can't connect my phone to it because of security reasons. @tofu951753 If you are having the same issue(native FGA screen recorder not working), would you be able to extract the crash logs?

**Comment by reconman** [comment]
Just guessing, but I think it doesn't affect tofu because his FGA can still detect stuff on the screen.

You can try https://play.google.com/store/apps/details?id=com.dp.logcatapp&hl=de&gl=US, but I have no idea if it works.

**Comment by ScathachSkadi** [comment]
It said it needed ADB at the start, but I ignored it to see if it still works(probably not)

https://pastebin.com/ADs2Yqdk

I will ask my friends if they are having the same issue.

**Comment by reconman** [comment]
Add some exclusions please from the ...options in the top right.

You have to create one per unwanted log level:
![Screenshot_20220226-115230.jpg](https://user-images.githubusercontent.com/12360257/155840523-1994b27f-3033-40a2-9fec-85ef56591c7a.jpg)

**Comment by ScathachSkadi** [comment]
I guess it didn't detect it then. To the FGO community I go

**Comment by reconman** [comment]
Do any errors appear if you produce the crash again?

**Comment by ScathachSkadi** [comment]
Nope. It's either the app crashing without any notice or app crashing and the system saying that it keeps crashing.

**Comment by tofu951753** [comment]
@reconman @ScathachSkadi 
It's working fine for me now. There was another software update for me about a week ago and it fixed the issues. Probably a bug in the previous firmware update that affected screen detection apps. 

Also, yes, you can't screen record with FGA running. There is some conflict with the screen recording software (I tried using the built in samsung game booster software for screen recording) that makes it so FGA can't detect the screen at all if you record.

**Comment by reconman** [comment]
@tofu951753 Their problem is that FGA crashes when recording with the correct recorder, as far as I understand.

**Comment by ScathachSkadi** [comment]
One UI 4.1 update just dropped and it still crashes. I got the logcat app to work and what filter/exclusions should I add @reconman ?

**Comment by reconman** [comment]
Should probably be enough to search for "mathew" in the top left. I'm achieving worse results with exclusions.

**Comment by ScathachSkadi** [comment]
https://pastebin.com/JX7fNMRG

I guess it takes time to show up on the logcat because it is only showing the ones from a few hours ago.

**Comment by ScathachSkadi** [comment]
https://file.io/RZ0A5B9VS0Di

I reopened the app and here it is

**Comment by reconman** [comment]
I see something about Accessibility screen readers being disabled:

`isScreenReaderEnable : false`

I think that's the part that's wrong.

Do you see anything related to that in Samsung's accessibility settings?

**Comment by ScathachSkadi** [comment]
I have not enabled or disabled or even added a new accessibility related app for ages. I didn't see a new one after the update either.

**Comment by reconman** [comment]
There's a new build available at https://github.com/Fate-Grand-Automata/FGA/suites/5817185474/artifacts/194854346 which should at least fix the Toaster error message in the logs.

Can you install the fix and then reproduce the bug?

The other error log is about Accessibility Service not running, which also happens to me from time to time, usually in the morning. I've already disabled energy optimization and locked the app in the task manager but it still happens. Then I have to force-kill FGA and then restart the Accessibility service.

Regarding the `isScreenReaderEnable`, it may relate to the screen reader feature of Android 12, which reads all the text on the screen. Do you see any new configuration options near Accessibility?

**Comment by reconman** [comment]
Another build, the first one probably doesn't work: https://github.com/Fate-Grand-Automata/FGA/suites/5817716262/artifacts/194887810

**Comment by ScathachSkadi** [comment]
I couldn't reproduce the crash on the new build. It just says it can't record the screen. 

The app's battery optimization setting was on compatibility mode so I changed it to no limit and I locked it on task manager as well on Galaxy Labs.... And I think it is working now. Nothing new on Accessibility but I believe it's the battery optimization. I can't believe I missed that. I think they changed how the battery optimization on Android 12. I swear it was on no limit before. Anyway it solved the problem for me... Thank you for the help.
